# Byte BPE (BBPE) Tutorial
## Ye Kyaw Thu, Lab Leader, Language Understanding Lab, Myanmar
## Date: 26 July 2026

ဒီtutorial မှာ minbpe, tokenizers နဲ့ tiktoken သုံးမျိုးကို မိတ်ဆက်ပေးရင်း Byte Pair Encoding (BPE) အခြေခံတဲ့ sub-word tokenization လုပ်တာကို လက်တွေ့လုပ်ပြသွားပါမယ်။ လက်ရှိ Large Language Model (LLM) တွေမှာ အဓိကသုံးတဲ့ sub-word segmentation သို့မဟုတ် tokenization နည်းလမ်းတွေမို့ AI Engineering ကျောင်းသားတိုင်းက သုံးတတ်ဖို့ လိုအပ်ပါတယ်။  

## minbpe

Link: https://github.com/karpathy/minbpe  

BBPE ကိုနားလည်လွယ်အောင် ရေးထားတဲ့ coding ပါ။ Educational purpose အတွက်ပါ။  
ကိုယ့်စက်ထဲမှာ သုံးလို့ရအောင် အရင်ဆုံး installation လုပ်ကြရအောင်။  


In [27]:
%cd /home/ye/aif2/word-seg/tool

/home/ye/aif2/word-seg/tool


In [10]:
!git clone https://github.com/karpathy/minbpe

Cloning into 'minbpe'...
remote: Enumerating objects: 217, done.
remote: Counting objects: 100% (123/123), done.
remote: Compressing objects: 100% (32/32), done.
remote: Total 217 (delta 104), reused 91 (delta 91), pack-reused 94 (from 2)
Receiving objects: 100% (217/217), 331.84 KiB | 3.39 MiB/s, done.
Resolving deltas: 100% (131/131), done.


In [28]:
%cd minbpe/

/home/ye/aif2/word-seg/tool/minbpe


In [18]:
!ls

assets	     lecture.md  minbpe     requirements.txt  train.py
exercise.md  LICENSE	 README.md  tests


In [19]:
!cat requirements.txt

regex
tiktoken

**--break-system-packages က လိုအပ်မှသုံးပါ။**

In [34]:
!pip install -r requirements.txt --break-system-packages

Defaulting to user installation because normal site-packages is not writeable
  Using cached tiktoken-0.13.0-cp312-cp312-manylinux_2_28_x86_64.whl.metadata (6.7 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 801.8/801.8 kB 5.9 MB/s  0:00:00
Using cached tiktoken-0.13.0-cp312-cp312-manylinux_2_28_x86_64.whl (1.1 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [tiktoken]
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
sentence-transformers 5.0.0 requires transformers<5.0.0,>=4.41.0, but you have transformers 5.7.0 which is incompatible.


## Train minbpe Tokenizer

In [38]:
import os
import time


# Tell Python where to find the cloned minbpe folder
sys.path.append(os.path.abspath("./minbpe"))

from minbpe import BasicTokenizer, RegexTokenizer

# open some text and train a vocab of 512 tokens
text = open("/home/ye/aif2/word-seg/data/myPOS/otest.txt", "r", encoding="utf-8").read()

# create a directory for models, so we don't pollute the current directory
os.makedirs("models", exist_ok=True)

t0 = time.time()
for TokenizerClass, name in zip([BasicTokenizer, RegexTokenizer], ["basic", "regex"]):

    # construct the Tokenizer object and kick off verbose training
    tokenizer = TokenizerClass()
    tokenizer.train(text, 512, verbose=True)
    # writes two files in the models directory: name.model, and name.vocab
    prefix = os.path.join("models", name)
    tokenizer.save(prefix)
t1 = time.time()

print(f"Training took {t1 - t0:.2f} seconds")


merge 1/256: (225, 128) -> 256 (b'\xe1\x80') had 53771 occurrences
merge 2/256: (32, 256) -> 257 (b' \xe1\x80') had 10846 occurrences
merge 3/256: (256, 186) -> 258 (b'\xe1\x80\xba') had 6569 occurrences
merge 4/256: (256, 172) -> 259 (b'\xe1\x80\xac') had 3459 occurrences
merge 5/256: (256, 184) -> 260 (b'\xe1\x80\xb8') had 3269 occurrences
merge 6/256: (258, 257) -> 261 (b'\xe1\x80\xba \xe1\x80') had 2836 occurrences
merge 7/256: (256, 175) -> 262 (b'\xe1\x80\xaf') had 2508 occurrences
merge 8/256: (256, 177) -> 263 (b'\xe1\x80\xb1') had 2396 occurrences
merge 9/256: (256, 173) -> 264 (b'\xe1\x80\xad') had 2187 occurrences
merge 10/256: (260, 257) -> 265 (b'\xe1\x80\xb8 \xe1\x80') had 2123 occurrences
merge 11/256: (258, 256) -> 266 (b'\xe1\x80\xba\xe1\x80') had 2063 occurrences
merge 12/256: (256, 132) -> 267 (b'\xe1\x80\x84') had 1863 occurrences
merge 13/256: (225, 129) -> 268 (b'\xe1\x81') had 1790 occurrences
merge 14/256: (264, 262) -> 269 (b'\xe1\x80\xad\xe1\x80\xaf') had 1596

## Check models Folder

In [39]:
%ls models

basic.model  basic.vocab  regex.model  regex.vocab


In [40]:
!wc ./models/*.vocab

  512  1787 11267 ./models/basic.vocab
  512  1687  9397 ./models/regex.vocab
 1024  3474 20664 total


In [41]:
!head -n 20 ./models/basic.vocab

[\u0000] 0
[\u0001] 1
[\u0002] 2
[\u0003] 3
[\u0004] 4
[\u0005] 5
[\u0006] 6
[\u0007] 7
[\u0008] 8
[\u0009] 9
[\u000a] 10
[\u000b] 11
[\u000c] 12
[\u000d] 13
[\u000e] 14
[\u000f] 15
[\u0010] 16
[\u0011] 17
[\u0012] 18
[\u0013] 19


In [42]:
!head -n 20 ./models/regex.vocab

[\u0000] 0
[\u0001] 1
[\u0002] 2
[\u0003] 3
[\u0004] 4
[\u0005] 5
[\u0006] 6
[\u0007] 7
[\u0008] 8
[\u0009] 9
[\u000a] 10
[\u000b] 11
[\u000c] 12
[\u000d] 13
[\u000e] 14
[\u000f] 15
[\u0010] 16
[\u0011] 17
[\u0012] 18
[\u0013] 19


In [45]:
!mv models ./tmp/

In [46]:
!tree ./tmp/

./tmp/
└── models
    ├── basic.model
    ├── basic.vocab
    ├── regex.model
    └── regex.vocab

2 directories, 4 files


## Update the Code for Showing Myanmar Characters and Retrain minbpe Tokenizer

In [50]:
import os
import time
import sys

# Tell Python where to find the cloned minbpe folder
sys.path.append(os.path.abspath("./minbpe"))

from minbpe import BasicTokenizer, RegexTokenizer

# Open your Myanmar text file
text = open("/home/ye/aif2/word-seg/data/myPOS/otest.txt", "r", encoding="utf-8").read()

# Create a directory for models
os.makedirs("models", exist_ok=True)

t0 = time.time()
for TokenizerClass, name in zip([BasicTokenizer, RegexTokenizer], ["basic", "regex"]):
    print(f"\n--- Training {name.upper()} Tokenizer on Myanmar Text ---")
    
    tokenizer = TokenizerClass()
    
    # Custom training loop or inspecting merges:
    # minbpe's tokenizer.train prints internal prints natively, but we can 
    # also override/monkey-patch or decode the final vocab/merges to show my students.
    tokenizer.train(text, 512, verbose=True)
    
    prefix = os.path.join("models", name)
    tokenizer.save(prefix)

t1 = time.time()
print(f"\nTraining took {t1 - t0:.2f} seconds")

# --- DEMONSTRATION: Print ALL Learned Merges Human-Readably ---
#print(f"\n--- Total Merges Learned: {len(tokenizer.merges)} ---")

#for i, (pair, new_token_id) in enumerate(tokenizer.merges.items()):
#    p1, p2 = pair
#    merged_bytes = tokenizer.vocab[new_token_id]
    
    # Safely decode the bytes to text
#    try:
#        readable_str = merged_bytes.decode("utf-8", errors="strict")
#        display_str = f"'{readable_str}'"
#    except UnicodeDecodeError:
        # For partial multi-byte UTF-8 sequences (like building blocks of a letter)
#        display_str = f"[Partial: {merged_bytes.decode('utf-8', errors='replace')}]"
    
#    print(f"Merge {i+1}/{len(tokenizer.merges)}: tokens ({p1}, {p2}) -> ID {new_token_id} -> {display_str}")

# --- CLEAN PRINT FORMAT FOR MYANMAR BPE MERGES ---
print(f"\n--- Total Merges Learned: {len(tokenizer.merges)} ---")

for i, (pair, new_token_id) in enumerate(tokenizer.merges.items()):
    p1, p2 = pair
    merged_bytes = tokenizer.vocab[new_token_id]
    
    # Safely convert bytes to a readable string representation
    try:
        # Try full strict decoding
        decoded_str = merged_bytes.decode("utf-8", errors="strict")
        display_str = repr(decoded_str)  # Adds quotes like 'က' or ' တ'
    except UnicodeDecodeError:
        # If it's a partial multi-byte fragment, show it as a hex-safe byte representation
        display_str = f"[Partial bytes: {merged_bytes}]"
    
    print(f"merge {i+1}/256: tokens {pair} -> ID {new_token_id} ({display_str})")


--- Training BASIC Tokenizer on Myanmar Text ---
merge 1/256: (225, 128) -> 256 (b'\xe1\x80') had 53771 occurrences
merge 2/256: (32, 256) -> 257 (b' \xe1\x80') had 10846 occurrences
merge 3/256: (256, 186) -> 258 (b'\xe1\x80\xba') had 6569 occurrences
merge 4/256: (256, 172) -> 259 (b'\xe1\x80\xac') had 3459 occurrences
merge 5/256: (256, 184) -> 260 (b'\xe1\x80\xb8') had 3269 occurrences
merge 6/256: (258, 257) -> 261 (b'\xe1\x80\xba \xe1\x80') had 2836 occurrences
merge 7/256: (256, 175) -> 262 (b'\xe1\x80\xaf') had 2508 occurrences
merge 8/256: (256, 177) -> 263 (b'\xe1\x80\xb1') had 2396 occurrences
merge 9/256: (256, 173) -> 264 (b'\xe1\x80\xad') had 2187 occurrences
merge 10/256: (260, 257) -> 265 (b'\xe1\x80\xb8 \xe1\x80') had 2123 occurrences
merge 11/256: (258, 256) -> 266 (b'\xe1\x80\xba\xe1\x80') had 2063 occurrences
merge 12/256: (256, 132) -> 267 (b'\xe1\x80\x84') had 1863 occurrences
merge 13/256: (225, 129) -> 268 (b'\xe1\x81') had 1790 occurrences
merge 14/256: (264, 

## UTF-8 and Multi-Byte Characters

အင်္ဂလိပ်စာမှာက letter တိုင်းက single byte (i.e. ASCII 0-127) နဲ့ အဆင်ပြေတယ်။ ဥပမာ 'e' က byte '101'။  
ဒါပေမဲ့ UTF-8 မြန်မာစာမှာက character တစ်လုံးကို 3 bytes နဲ့ ကိုယ်စားပြုတယ်။ မြန်မာစာလုံးတိုင်းက prefix bytes '\xe1\x80' နဲ့စတယ်။  

## Step-by-Step Line Explanation

**Merge 1:** tokens (225, 128) -> ID 256 ([Partial bytes: b'\xe1\x80'])
Algorithm က byte 225 (\xe1) နဲ့ byte 128 (\x80) ကို တွဲလျက် ခဏခဏတွေ့ရတယ်။ တခြား အတွဲတွေထက်ကို ဒီအတွဲကို ပိုတွေ့ရတာမို့ အရင်ဆုံး စပြီး merge လုပ်တယ်။ တနည်းအားဖြင့် BPE စလုပ်ပါတယ်။  

**This is the foundational building block for every Myanmar letter.**  
ဒီ ID 256 က မြန်မာစာရဲ့ ဗျည်းတို့ သရတို့ စတဲ့ စာလုံးတွေကို ကိုယ်စားပြုတာ မဟုတ်ပေမဲ့၊ မြန်မာစာစာလုံးတိုင်းမှာ ရှေ့ကနေ ပါနေတဲ့ 2-byte prefix (\xe1\x80) ကို ကိုယ်စားပြုပါတယ်။ 

**Merge 2:** tokens (32, 256) -> ID 257 ([Partial bytes: b' \xe1\x80'])
ဒီmerge ကကျတော့ ID 256 (the prefix \xe1\x80) နဲ့Byte 32 (a space ' ') အတွဲအတွက် BPE လုပ်သွားတာပါ။  

**This shows how BPE handles word boundaries.**   
ဒါကဘာကိုဆိုလိုတာလဲ ဆိုတော့ algorithm က မြန်မာစာစာလုံး သို့မဟုတ် ဝဏ္ဏတွေရဲ့ ရှေ့မှာ မကြာခဏ space ရှိတတ်တယ်ဆိုတာကို ဂရုပြုမိတယ်လို့ နားလည်လို့ ရပါတယ်။  

*Merges 3 to 8: ဒီနေရာကနေ စပြီး ပုံမှန်မြန်မာစာ စာလုံးတွေဖြစ်တဲ့ သရတို့ အသတ်တို့ကို စတင်မြင်ရပါပြီ။  

**Merge 3:** (256, 186) -> ID 258 ('်')

**Merge 4:** (256, 172) -> ID 259 ('ာ')

**Merge 5:** (256, 184) -> ID 260 ('း')

**Merge 6:** (256, 175) -> ID 261 ('ု')

**Merge 7:** (256, 177) -> ID 262 ('ေ')

**Merge 8:** (256, 173) -> ID 263 ('ိ') 

In [62]:
%pwd

'/home/ye/aif2/word-seg/tool/minbpe'

In [63]:
%cd models

/home/ye/aif2/word-seg/tool/minbpe/models


In [64]:
!ls * -lh

-rw-rw-r-- 1 ye ye 1.9K Jul 26 11:34 basic.model
-rw-rw-r-- 1 ye ye 8.1K Jul 26 11:34 basic.vocab
-rw-rw-r-- 1 ye ye 2.0K Jul 26 11:34 regex.model
-rw-rw-r-- 1 ye ye 7.7K Jul 26 11:34 regex.vocab


In [65]:
!wc basic.vocab

 512 1809 8276 basic.vocab


In [66]:
!head -n 50 ./basic.vocab

[\u0000] 0
[\u0001] 1
[\u0002] 2
[\u0003] 3
[\u0004] 4
[\u0005] 5
[\u0006] 6
[\u0007] 7
[\u0008] 8
[\u0009] 9
[\u000a] 10
[\u000b] 11
[\u000c] 12
[\u000d] 13
[\u000e] 14
[\u000f] 15
[\u0010] 16
[\u0011] 17
[\u0012] 18
[\u0013] 19
[\u0014] 20
[\u0015] 21
[\u0016] 22
[\u0017] 23
[\u0018] 24
[\u0019] 25
[\u001a] 26
[\u001b] 27
[\u001c] 28
[\u001d] 29
[\u001e] 30
[\u001f] 31
[ ] 32
[!] 33
["] 34
[#] 35
[$] 36
[%] 37
[&] 38
['] 39
[(] 40
[)] 41
[*] 42
[+] 43
[,] 44
[-] 45
[.] 46
[/] 47
[0] 48
[1] 49


In [67]:
!tail -n 50 ./basic.vocab

[ep][t] -> [ept] 462
[Ma][y ] -> [May ] 463
[, 201][5] -> [, 2015] 464
[Ap][ri] -> [Apri] 465
[Apri][l ] -> [April ] 466
[l][e ] -> [le ] 467
[A][w] -> [Aw] 468
[ati][on] -> [ation] 469
[S][ept] -> [Sept] 470
[Sept][ember ] -> [September ] 471
[r][a] -> [ra] 472
[al][bum] -> [album] 473
[C][h] -> [Ch] 474
[v][e ] -> [ve ] 475
[es][t ] -> [est ] 476
[J][an] -> [Jan] 477
[2][2] -> [22] 478
[Jan][uary ] -> [January ] 479
[oun][tr] -> [ountr] 480
[ig][h] -> [igh] 481
[". ][The ] -> [". The ] 482
[, 2023][.\u000a ] -> [, 2023.\u000a ] 483
[1][3] -> [13] 484
[A][l] -> [Al] 485
[e][t] -> [et] 486
[es][s] -> [ess] 487
[Mar][ch ] -> [March ] 488
[u][t] -> [ut] 489
[w][rit] -> [writ] 490
[l][o] -> [lo] 491
[s][ong] -> [song] 492
[�][�] -> [�] 493
[ar][d ] -> [ard ] 494
[0][ ] -> [0 ] 495
[u][l] -> [ul] 496
[2][4] -> [24] 497
[i][s ] -> [is ] 498
[ti][c] -> [tic] 499
[a][g] -> [ag] 500
["][ ] -> [" ] 501
[A][n] -> [An] 502
[1][8] -> [18] 503
[f][or] -> [for] 504
[ountr][y ] -> [ountry ] 505
[A][m

In [68]:
!head -n 50 ./regex.vocab

[\u0000] 0
[\u0001] 1
[\u0002] 2
[\u0003] 3
[\u0004] 4
[\u0005] 5
[\u0006] 6
[\u0007] 7
[\u0008] 8
[\u0009] 9
[\u000a] 10
[\u000b] 11
[\u000c] 12
[\u000d] 13
[\u000e] 14
[\u000f] 15
[\u0010] 16
[\u0011] 17
[\u0012] 18
[\u0013] 19
[\u0014] 20
[\u0015] 21
[\u0016] 22
[\u0017] 23
[\u0018] 24
[\u0019] 25
[\u001a] 26
[\u001b] 27
[\u001c] 28
[\u001d] 29
[\u001e] 30
[\u001f] 31
[ ] 32
[!] 33
["] 34
[#] 35
[$] 36
[%] 37
[&] 38
['] 39
[(] 40
[)] 41
[*] 42
[+] 43
[,] 44
[-] 45
[.] 46
[/] 47
[0] 48
[1] 49


In [69]:
!tail -n 50 ./basic.vocab

[ep][t] -> [ept] 462
[Ma][y ] -> [May ] 463
[, 201][5] -> [, 2015] 464
[Ap][ri] -> [Apri] 465
[Apri][l ] -> [April ] 466
[l][e ] -> [le ] 467
[A][w] -> [Aw] 468
[ati][on] -> [ation] 469
[S][ept] -> [Sept] 470
[Sept][ember ] -> [September ] 471
[r][a] -> [ra] 472
[al][bum] -> [album] 473
[C][h] -> [Ch] 474
[v][e ] -> [ve ] 475
[es][t ] -> [est ] 476
[J][an] -> [Jan] 477
[2][2] -> [22] 478
[Jan][uary ] -> [January ] 479
[oun][tr] -> [ountr] 480
[ig][h] -> [igh] 481
[". ][The ] -> [". The ] 482
[, 2023][.\u000a ] -> [, 2023.\u000a ] 483
[1][3] -> [13] 484
[A][l] -> [Al] 485
[e][t] -> [et] 486
[es][s] -> [ess] 487
[Mar][ch ] -> [March ] 488
[u][t] -> [ut] 489
[w][rit] -> [writ] 490
[l][o] -> [lo] 491
[s][ong] -> [song] 492
[�][�] -> [�] 493
[ar][d ] -> [ard ] 494
[0][ ] -> [0 ] 495
[u][l] -> [ul] 496
[2][4] -> [24] 497
[i][s ] -> [is ] 498
[ti][c] -> [tic] 499
[a][g] -> [ag] 500
["][ ] -> [" ] 501
[A][n] -> [An] 502
[1][8] -> [18] 503
[f][or] -> [for] 504
[ountr][y ] -> [ountry ] 505
[A][m

## Let's Check with diff Command

In [70]:
!diff ./basic.vocab ./regex.vocab | head -n 50

257,290c257,290
< [e][ ] -> [e ] 256
< [,][ ] -> [, ] 257
< [d][ ] -> [d ] 258
< [.][ ] -> [. ] 259
< [r][ ] -> [r ] 260
< [2][0] -> [20] 261
< [s][ ] -> [s ] 262
< [i][n] -> [in] 263
< [o][n] -> [on] 264
< [r][i] -> [ri] 265
< [t][ ] -> [t ] 266
< [t][h] -> [th] 267
< [e][d ] -> [ed ] 268
< [, ][20] -> [, 20] 269
< [a][n] -> [an] 270
< [a][r] -> [ar] 271
< [e][r ] -> [er ] 272
< [y][ ] -> [y ] 273
< [a][l] -> [al] 274
< [th][e ] -> [the ] 275
< [v][ed ] -> [ved ] 276
< [w][i] -> [wi] 277
< [e][r] -> [er] 278
< [on][ ] -> [on ] 279
< [wi][f] -> [wif] 280
< [R][e] -> [Re] 281
< [S][wif] -> [Swif] 282
< [o][r ] -> [or ] 283
< [c][h] -> [ch] 284
< [, 20][1] -> [, 201] 285
< [o][m] -> [om] 286
< [b][er ] -> [ber ] 287
< [ ][the ] -> [ the ] 288
< [a][y] -> [ay] 289
---
> [e][r] -> [er] 256
> [2][0] -> [20] 257
> [o][r] -> [or] 258
> [i][n] -> [in] 259
> [e][d] -> [ed] 260
> [ ][t] -> [ t] 261
> [o][n] -> [on] 262
> [h][e] -> [he] 263
> [ ][S] -> [ S] 264
> [a][r] -> [ar] 265
> [a][n] -> [a

## Testing minbpe Model

In [52]:
import os
import sys

# Ensure Python knows where to find the minbpe library
sys.path.append(os.path.abspath("./minbpe"))

from minbpe import BasicTokenizer, RegexTokenizer

# --- CHOOSE WHICH MODEL TO LOAD ---
# You can change this to "regex" if you want to test the RegexTokenizer model
model_type = "basic" 
prefix = os.path.join("models", model_type)

# 1. Instantiate the appropriate Tokenizer class
if model_type == "regex":
    tokenizer = RegexTokenizer()
else:
    tokenizer = BasicTokenizer()

# 2. Load the trained model files (this looks for models/basic.model and models/basic.vocab)
tokenizer.load(prefix + ".model")
print(f"Successfully loaded '{model_type}' tokenizer! Vocabulary size: {len(tokenizer.vocab)}")

# 3. Test encoding and decoding
test_text = "အိန္ဒိယကျောင်းသားလှုပ်ရှားမှု"
encoded_tokens = tokenizer.encode(test_text)
print("\nEncoded IDs:", encoded_tokens)

decoded_text = tokenizer.decode(encoded_tokens)
print("Decoded Text:", decoded_text)

# Optional: Verify it's a lossless round-trip
assert test_text == decoded_text, "Decoded text does not match original!"
print("Round-trip test passed successfully!")

Successfully loaded 'basic' tokenizer! Vocabulary size: 512

Encoded IDs: [256, 161, 264, 279, 375, 256, 146, 264, 306, 350, 274, 372, 158, 366, 156, 290, 502, 324, 366, 299, 262]
Decoded Text: အိန္ဒိယကျောင်းသားလှုပ်ရှားမှု
Round-trip test passed successfully!


## Production Byte-Level BPE (BBPE) with Hugging Face 

Link: https://github.com/huggingface/tokenizers   
အရင်ဆုံး installation လုပ်ကြရအောင်။  

In [54]:
!pip install tokenizers --break-system-packages

Defaulting to user installation because normal site-packages is not writeable


In [55]:
%pwd

'/home/ye/aif2/word-seg/tool/minbpe'

In [57]:
from tokenizers import Tokenizer, decoders
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import ByteLevel

# 1. Initialize a Byte-Level BPE model
tokenizer = Tokenizer(BPE(unk_token="<unk>"))

# 2. IMPORTANT: Set BOTH the Pre-Tokenizer AND the Decoder for Byte-Level mapping
tokenizer.pre_tokenizer = ByteLevel(add_prefix_space=False)
tokenizer.decoder = decoders.ByteLevel()

# 3. Setup the trainer with a target vocabulary size
trainer = BpeTrainer(
    vocab_size=1000, 
    special_tokens=["<unk>"]
)

# 4. Train on your Myanmar text file
files = ["/home/ye/aif2/word-seg/sub-word/big_data/bbc-cockroach-news.txt"]
tokenizer.train(files, trainer)

# 5. Encode text and look at the behavior
output = tokenizer.encode("အိန္ဒိယကျောင်းသားလှုပ်ရှားမှု")
print("Tokens:", output.tokens)
print("IDs:", output.ids)

# 6. Decode back using the proper decoder
decoded_text = tokenizer.decode(output.ids)
print("Decoded Text:", decoded_text)




Tokens: ['áĢ¡', 'áĢŃ', 'áĢĶ', 'áĢ¹', 'áĢĴ', 'áĢŃ', 'áĢļ', 'áĢĢ', 'áĢ»áĢ±áĢ¬', 'áĢĦ', 'áĢºáĢ¸', 'áĢŀ', 'áĢ¬áĢ¸', 'áĢľ', 'áĢ¾áĢ¯', 'áĢķ', 'áĢº', 'áĢĽ', 'áĢ¾áĢ¬áĢ¸', 'áĢĻ', 'áĢ¾áĢ¯']
IDs: [109, 83, 88, 118, 112, 83, 101, 81, 153, 84, 96, 107, 97, 103, 105, 82, 74, 90, 127, 85, 105]
Decoded Text: အိန္ဒိယကျောင်းသားလှုပ်ရှားမှု


## For Your Information

Hugging Face's ByteLevel pre-tokenizer uses a clever trick invented for GPT-2: it maps raw bytes (0–255) to a specific set of visible, printable Unicode characters so that control characters, spaces, and binary data don't break string formatting or tokenization files.  

A space ( ) is mapped to the character Ġ (U+0120).  

Raw bytes of Myanmar letters (like \xe1\x80) are mapped to printable ASCII/Latin equivalents like áĢ¡ during the intermediate tokenization steps.  

## High-Speed Production Tokenization with tiktoken

Demonstrate how modern LLMs (like GPT-4) use fixed pre-trained.  

In [59]:
!pip install tiktoken --break-system-packages

Defaulting to user installation because normal site-packages is not writeable


In [60]:
import tiktoken

# OpenAI doesn't natively have a pre-trained Myanmar-only tiktoken encoder out-of-the-box, 
# but we can inspect how cl100k_base (GPT-4 encoding) handles Myanmar text via bytes.
enc = tiktoken.get_encoding("cl100k_base")

text_sample = "အိန္ဒိယနိုင်ငံမှာ အကြီးမားဆုံး ကျောင်းသားလှုပ်ရှားမှု"
encoded = enc.encode(text_sample)

print("GPT-4 Token IDs for Myanmar text:", encoded)
print("Number of tokens:", len(encoded))

# Decode back to verify lossless round-trip
print("Decoded back:", enc.decode(encoded))

GPT-4 Token IDs for Myanmar text: [25870, 94, 25870, 255, 25870, 242, 25870, 117, 25870, 240, 25870, 255, 25870, 248, 25870, 242, 25870, 255, 25870, 107, 25870, 226, 25870, 118, 25870, 226, 25870, 114, 25870, 247, 25870, 122, 25870, 105, 220, 25870, 94, 25870, 222, 25870, 120, 25870, 106, 25870, 116, 25870, 247, 25870, 105, 25870, 116, 25870, 228, 25870, 107, 25870, 114, 25870, 116, 220, 25870, 222, 25870, 119, 25870, 109, 25870, 105, 25870, 226, 25870, 118, 25870, 116, 25870, 252, 25870, 105, 25870, 116, 25870, 250, 25870, 122, 25870, 107, 25870, 243, 25870, 118, 25870, 249, 25870, 122, 25870, 105, 25870, 116, 25870, 247, 25870, 122, 25870, 107]
Number of tokens: 104
Decoded back: အိန္ဒိယနိုင်ငံမှာ အကြီးမားဆုံး ကျောင်းသားလှုပ်ရှားမှု


## What is cl100k_base?  

- cl100k_base မော်ဒယ်က OpenAI က သူ့ရဲ့ မော်ဒယ်အများစုအတွက် သုံးနေတဲ့ pre-trained encoding မော်ဒယ်ပါပဲ။  
- cl (ChatGPT/Clrir) က GPT-4 မော်ဒယ်လောက်ကနေ စထွက်လာတဲ့ မော်ဒယ်တွေကို ညွှန်းပါတယ်။  
- base ဆိုတာက task-specific fine-tuning မတိုင်ခင် အခြေခံ layer ကို ဆိုလိုတာပါ။  
- GPT-4, GPT-4o, and GPT-3.5-Turbo တွေမှာ default tokenizer အဖြစ်သုံးတဲ့ မော်ဒယ်ပါ။  

## Summary

ဒီ tutorial ကနေ BBPE နဲ့ ပတ်သက်ပြီး တော်တော်လေး ဗဟုသုတရသွားမယ်လို့ ယုံကြည်ပါတယ်။  

| Feature | `minbpe` (Karpathy) | Hugging Face `tokenizers` | OpenAI `tiktoken` |
| :--- | :--- | :--- | :--- |
| **Primary Purpose** | Teaching & Code Readability | Production & Research | Ultra-fast Production (OpenAI models) |
| **Under the Hood** | Pure Python | Rust (with Python bindings) | Rust (with Python bindings) |
| **Speed** | Slow (good for small text/demos) | Extremely Fast | Extremely Fast |
| **Custom Training** | Excellent (you see every merge step) | Excellent (customizable pipelines) | Supports custom training via API |